zkvm-host/bin/agg.rs

```diff
@@ -0,0 +1,119 @@
+use std::fs;
+
+use anyhow::Result;
+use cargo_metadata::MetadataCommand;
+use clap::Parser;
+use client_utils::{RawBootInfo, BOOT_INFO_SIZE};
+use host_utils::fetcher::{ChainMode, SP1KonaDataFetcher};
+use sp1_sdk::{utils, ProverClient, SP1Proof, SP1ProofWithPublicValues, SP1Stdin};
+
+pub const AGG_ELF: &[u8] = include_bytes!("../../elf/aggregation-client-elf");
+pub const MULTI_BLOCK_ELF: &[u8] = include_bytes!("../../elf/validity-client-elf");
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Start L2 block number.
+    #[arg(short, long, num_args = 1.., value_delimiter = ',')]
+    proofs: Vec<String>,
+
+    #[arg(short, long)]
+    multi: bool,
+
+    /// Verbosity level.
+    #[arg(short, long, default_value = "0")]
+    verbosity: u8,
+}
+
+/// Load the aggregation proof data.
+fn load_aggregation_proof_data(proof_names: Vec<String>) -> (Vec<SP1Proof>, Vec<RawBootInfo>) {
+    let metadata = MetadataCommand::new().exec().unwrap();
+    let workspace_root = metadata.workspace_root;
+    let proof_directory = format!("{}/data/proofs", workspace_root);
+
+    let mut proofs = Vec::with_capacity(proof_names.len());
+    let mut boot_infos = Vec::with_capacity(proof_names.len());
+
+    for proof_name in proof_names.iter() {
+        let proof_path = format!("{}/{}.bin", proof_directory, proof_name);
+        if fs::metadata(&proof_path).is_err() {
+            panic!("Proof file not found: {}", proof_path);
+        }
+        let mut deserialized_proof =
+            SP1ProofWithPublicValues::load(proof_path).expect("loading proof failed");
+        proofs.push(deserialized_proof.proof);
+
+        // The public values are the ABI-encoded BootInfo.
+        let mut boot_info_buf = [0u8; BOOT_INFO_SIZE];
+        deserialized_proof
+            .public_values
+            .read_slice(&mut boot_info_buf);
+        let boot_info = RawBootInfo::abi_decode(&boot_info_buf).unwrap();
+        boot_infos.push(boot_info);
+    }
+
+    (proofs, boot_infos)
+}
+
+// Execute the Kona program for a single block.
+#[tokio::main]
+async fn main() -> Result<()> {
+    utils::setup_logger();
+
+    let args = Args::parse();
+    let fetcher = SP1KonaDataFetcher::new();
+    let prover = ProverClient::new();
+
+    let (proofs, boot_infos) = load_aggregation_proof_data(args.proofs);
+
+    // Fetch the headers from the L1 head of the last block to the L1 head of the first block.
+    let first_head = boot_infos.last().unwrap().l1_head;
```

Comment:
Instead of defaulting to starting at the last `boot_infos` L1 Head, it would be great if we could pass an additional L1 Head as a CLI argument.

This would make it way easier on the proposer side, because we need to checkpoint whatever L1 Head we're going to use for this proof, and we won't know to do that until we know we're ready to do the aggregation proof.

And a later header will still be able to be walked back to everything we need.